File: superstitioApi/java/superstitioapi/renderManager/inBattleManager/InBattleDataManager.java
package superstitioapi.renderManager.inBattleManager;

import basemod.interfaces.ISubscriber;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import superstitioapi.actions.AutoDoneInstantAction;
import superstitioapi.hangUpCard.HangUpCardGroup;
import superstitioapi.pet.PetManager;
import superstitioapi.powers.barIndepend.BarRenderManager;

import java.util.*;
import java.util.function.Consumer;

public class InBattleDataManager {
    public static Map<UUID, Integer> costMap = new HashMap<>();
    public static ArrayList<ISubscriber> subscribeManageGroups = new ArrayList<>();

    public static void InitializeAtStartOfBattle() {
        ResetAll();
        AutoDoneInstantAction.addToTopAbstract(() ->
        {
            Subscribe(new HangUpCardGroup(AbstractDungeon.player.hb));
            Subscribe(new BarRenderManager());
            Subscribe(new PetManager());
        });
    }


    public static void ClearOnEndOfBattle() {
        ResetAll();
    }

    private static void ResetAll() {
        RenderInBattle.clearAll();
        costMap = new HashMap<>();
        subscribeManageGroups.clear();
    }

    public static <T extends ISubscriber> void ApplyAll(Consumer<T> subscriberConsumer, Class<T> tClass) {
        subscribeManageGroups.stream()
                .filter(tClass::isInstance)
                .map(object -> (T) object)
                .forEach(subscriberConsumer);
    }

//    public static <T extends ISubscriber> void ApplyAllForPower(Consumer<T> subscriberConsumer, Class<T> tClass) {
//        Manager.stream()
//                .filter(tClass::isInstance)
//                .map(object -> (T) object)
//                .forEach(subscriberConsumer);
//    }


    public static void InitializeAtStartOfTurn() {
    }

    @SafeVarargs
    public static <T extends ISubscriber> void Subscribe(T... objectToAdd) {
        subscribeManageGroups.addAll(Arrays.asList(objectToAdd));
    }

    @SafeVarargs
    public static <T extends ISubscriber> void UnSubscribe(T... objectToAdd) {
        subscribeManageGroups.removeAll(Arrays.asList(objectToAdd));
    }

    public static Optional<HangUpCardGroup> getHangUpCardOrbGroup() {
        return InBattleDataManager.subscribeManageGroups.stream()
                .filter(orbGroup -> orbGroup instanceof HangUpCardGroup)
                .map(orbGroup -> (HangUpCardGroup) orbGroup).findAny();
    }

    public static Optional<BarRenderManager> getBarRenderManager() {
        return InBattleDataManager.subscribeManageGroups.stream()
                .filter(barGroup -> barGroup instanceof BarRenderManager)
                .map(barGroup -> (BarRenderManager) barGroup).findAny();
    }

    public static Optional<PetManager> getPetManager() {
        return InBattleDataManager.subscribeManageGroups.stream()
                .filter(petGroup -> petGroup instanceof PetManager)
                .map(petGroup -> (PetManager) petGroup).findAny();
    }
}

File: superstitioApi/java/superstitioapi/renderManager/inBattleManager/InBattleDataManager.kt
package superstitioapi.renderManager.inBattleManager

import basemod.interfaces.ISubscriber
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import superstitioapi.actions.AutoDoneInstantAction
import superstitioapi.hangUpCard.HangUpCardGroup
import superstitioapi.pet.PetManager
import superstitioapi.powers.barIndepend.BarRenderManager
import java.util.*
import java.util.function.Consumer

object InBattleDataManager
{
    var costMap: HashMap<UUID, Int> = HashMap()
    var subscribeManageGroups: MutableList<ISubscriber> = ArrayList()

    @JvmStatic
    fun InitializeAtStartOfBattle()
    {
        ResetAll()
        AutoDoneInstantAction.addToTopAbstract {
            Subscribe(HangUpCardGroup(AbstractDungeon.player.hb))
            Subscribe(BarRenderManager())
            Subscribe(PetManager())
        }
    }

    @JvmStatic
    fun ClearOnEndOfBattle()
    {
        ResetAll()
    }

    @JvmStatic
    fun <T : ISubscriber> ApplyAll(subscriberConsumer: Consumer<T>?, tClass: Class<T>)
    {
        subscribeManageGroups.filterIsInstance(tClass).forEach(subscriberConsumer)
    }

    @JvmStatic
    fun InitializeAtStartOfTurn()
    {
    }

    //    public static <T extends ISubscriber> void ApplyAllForPower(Consumer<T> subscriberConsumer, Class<T> tClass) {
    //        Manager.stream()
    //                .filter(tClass::isInstance)
    //                .map(object -> (T) object)
    //                .forEach(subscriberConsumer);
    //    }
    @SafeVarargs
    fun <T : ISubscriber> Subscribe(vararg objectToAdd: T)
    {
        subscribeManageGroups.addAll(objectToAdd)
    }

    @SafeVarargs
    fun <T : ISubscriber> UnSubscribe(vararg objectToAdd: T)
    {
        subscribeManageGroups.removeAll(objectToAdd.toSet())
    }

    fun getHangUpCardOrbGroup(): HangUpCardGroup?
    {
        return subscribeManageGroups
            .filterIsInstance<HangUpCardGroup>()
            .firstOrNull()
    }

    fun getBarRenderManager(): BarRenderManager? = subscribeManageGroups
        .filterIsInstance<BarRenderManager>()
        .firstOrNull()

    fun getPetManager(): PetManager? = subscribeManageGroups
        .filterIsInstance<PetManager>()
        .firstOrNull()

    private fun ResetAll()
    {
        RenderInBattle.clearAll()
        costMap = HashMap()
        subscribeManageGroups.clear()
    }
}
